server/coordinator/procedure/transfer_leader.go

```diff
@@ -0,0 +1,326 @@
+	if err != nil {
+		cancelEventWithLog(event, err, "get request from event")
+		return
+	}
+	ctx := request.ctx
+
+	closeShardRequest := eventdispatch.CloseShardRequest{
+		ShardID: uint32(request.shardID),
+	}
+	if err := request.dispatch.CloseShard(ctx, request.oldLeaderNodeName, closeShardRequest); err != nil {
+		cancelEventWithLog(event, err, "close shard failed", zap.Uint32("shardId", uint32(request.shardID)), zap.String("oldLeaderName", request.oldLeaderNodeName))
+		return
+	}
+}
+
+func transferLeaderOpenNewShardCallback(event *fsm.Event) {
+	request, err := getRequestFromEvent(event)
+	if err != nil {
+		cancelEventWithLog(event, err, "get request from event")
+		return
+	}
+	ctx := request.ctx
+
+	getNodeShardResult, err := request.cluster.GetNodeShards(ctx)
+	if err != nil {
+		cancelEventWithLog(event, err, "get node shards failed")
+		return
+	}
+	var preVersion uint64
+	for _, shardNodeWithVersion := range getNodeShardResult.NodeShards {
+		if request.shardID == shardNodeWithVersion.ShardNode.ID {
+			preVersion = shardNodeWithVersion.ShardInfo.Version
+		}
+	}
+
+	openShardRequest := eventdispatch.OpenShardRequest{
+		Shard: cluster.ShardInfo{ID: request.shardID, Role: storage.ShardRoleLeader, Version: preVersion + 1},
+	}
+
+	if err := request.dispatch.OpenShard(ctx, request.newLeaderNodeName, openShardRequest); err != nil {
+		cancelEventWithLog(event, err, "open shard failed", zap.Uint32("shardId", uint32(request.shardID)), zap.String("newLeaderNode", request.newLeaderNodeName))
+		return
+	}
+}
+
+func transferLeaderFinishCallback(event *fsm.Event) {
+	request, err := getRequestFromEvent(event)
+	if err != nil {
+		cancelEventWithLog(event, err, "get request from event")
+		return
+	}
+
+	log.Info("transfer leader finish", zap.Uint32("shardID", uint32(request.shardID)), zap.String("oldLeaderNode", request.oldLeaderNodeName), zap.String("newLeaderNode", request.newLeaderNodeName))
+}
+
+func (p *TransferLeaderProcedure) updateStateWithLock(state State) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	p.state = state
+}
+
+// TODO: Consider refactor meta procedure convertor function, encapsulate as a tool function.
+func (p *TransferLeaderProcedure) convertToMeta() (Meta, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	rawData := TransferLeaderProcedurePersistRawData{
+		ID:                p.id,
+		FsmState:          p.fsm.Current(),
+		ShardID:           p.shardID,
+		OldLeaderNodeName: p.oldLeaderNodeName,
+		NewLeaderNodeName: p.newLeaderNodeName,
+		State:             p.state,
+	}
+	rawDataBytes, err := json.Marshal(rawData)
+	if err != nil {
+		return Meta{}, ErrDecodeRawData.WithCausef("marshal raw data failed, procedureID:%v, err:%v", p.shardID, err)
+	}
+
+	meta := Meta{
+		ID:      p.id,
+		Typ:     TransferLeader,
+		State:   p.state,
+		RawData: rawDataBytes,
+	}
+
+	return meta, nil
+}
+
+func getRequestFromEvent(event *fsm.Event) (*TransferLeaderCallbackRequest, error) {
```

Comment:
Can this define as `func getRequestFromEvent(event *fsm.Event, data Interface{}) error`.